apiserver/facades/agent/upgrader/unitupgrader.go

```diff
@@ -204,33 +207,33 @@ func (u *UnitUpgraderAPI) Tools(ctx context.Context, args params.Entities) (para
 			continue
 		}
 		if u.authorizer.AuthOwner(tag) {
-			result.Results[i] = u.getMachineTools(tag)
+			result.Results[i] = u.getMachineTools(ctx, tag)
 		}
 	}
 	return result, nil
 }
 
-func (u *UnitUpgraderAPI) getAssignedMachine(tag names.Tag) (*state.Machine, error) {
+func (u *UnitUpgraderAPI) getAssignedMachine(ctx context.Context, tag names.Tag) (*state.Machine, error) {
 	// Check that we really have a unit tag.
 	switch tag := tag.(type) {
 	case names.UnitTag:
-		unit, err := u.st.Unit(tag.Id())
+		unitName, err := coreunit.NewName(tag.Id())
 		if err != nil {
 			return nil, apiservererrors.ErrPerm
 		}
-		id, err := unit.AssignedMachineId()
+		id, err := u.applicationService.GetUnitMachineName(ctx, unitName)
```

Comment:
It might be useful to check for a not found error here

Reply:
Added check for this.